File: Top 100 Codes/02-Working with Numbers/25-Finding number of integers which has exactly x divisors/Findingnumberofintegerswhichhasexactlyxdivisors.cpp
#include <iostream>
#include <cmath>
using namespace std;

int main(){
    int n, x;
    
    // Take user input for n and x
    cout << "Enter the upper bound (n): ";
    cin >> n;
    cout << "Enter the number of divisors (x): ";
    cin >> x;
    
    // Variable to count the numbers with exactly x divisors
    int count = 0;
    
    // Loop over each number from 1 to n
    for (int i = 1; i <= n; i++){
        // Variable to count the factors of the current number i
        int count_factors = 0;
        
        // Loop from 1 to the square root of i to check for factors
        for (int j = 1; j <= sqrt(i); j++){
            if (i % j == 0){  // j is a divisor of i
                // If j is not the square root of i, then there are two divisors: j and i/j
                if (i / j != j)
                    count_factors += 2;
                else  // When j is the square root, count only once
                    count_factors++;
            }
        }
        
        // If the current number has exactly x divisors, increment our count
        if (count_factors == x)
            count++;
    }
    
    // Output the result
    cout << "The number of integers between 1 and " << n 
         << " with exactly " << x << " divisors is: " << count << endl;
    
    return 0;
}

File: Top 100 Codes/02-Working with Numbers/25-Finding number of integers which has exactly x divisors/Findingnumberofintegerswhichhasexactlyxdivisors.py
import math

def count_divisors(num):
    """Return the number of divisors of num."""
    count = 0
    limit = int(math.sqrt(num))
    for i in range(1, limit + 1):
        if num % i == 0:
            # If i is the square root of num, count it only once
            if i == num // i:
                count += 1
            else:
                count += 2
    return count

# Take user input for n and x
n = int(input("Enter the upper bound (n): "))
x = int(input("Enter the number of divisors (x): "))

count_numbers = 0
# Check each number from 1 to n
for i in range(1, n + 1):
    if count_divisors(i) == x:
        count_numbers += 1

# Output the result
print(f"The number of integers between 1 and {n} with exactly {x} divisors is: {count_numbers}")
